circuits/web/http.py

```diff
@@ -65,20 +65,14 @@ def protocol(self):
 
     @property
     def scheme(self):
-        if not hasattr(self, "_server"):
-            return
         return "https" if self._server.secure else "http"
 
     @property
     def base(self):
-        if getattr(self, "uri", None) is None:
```

Comment:
If `self.uri` is `None` (ie before the `ready` handler ran), the line below `self.uri.utf8()` would error out if. I guess you still need to check if `self.uri` is `None` but you can remove the `getattr`